### ISSUE TITLE
Consider adding alternatives to cloud9

### ISSUE THREAD

**Issue by nt1m** [JanitorTechnology/dockerfiles#103]
It would be nice to have cloud9 alternatives like brackets/the koding editor/...

**Comment by jankeromnes** [comment]
Agreed, alternatives would be great for choice (other cloud IDEs, ways to access containers via SSH or local IDEs, etc).

The problem is that for Janitor we customize Cloud9 in order to automate many project workflows (e.g. the Reviews panel, the `janitor.json` scripts, etc) and for each alternative we add, we either commit to re-implement all automations, or to offer a less automated tool (decreasing the value of Janitor vs local tools).

Then again, it would be very valuable to have at least one alternative, just in case we encounter a problem that prevents us from using Cloud9 at all. So I'm +1 on the idea even though I'm not yet able to invest a lot of effort into this.

**Comment by ishitatsuyuki** [comment]
I think we should go in the direction of phasing out C9, mainly due to it being the only nonfree component in Janitor (the SDK is limited by the EULA).

A promising alternative is [Eclipse Che](https://www.eclipse.org/che/), which is very actively developed.

**Comment by jankeromnes** [comment]
Thanks a lot for the suggestions! I'm very interested in Eclipse Che, and I'd love to see how hard/easy it is to integrate it with Janitor.

However, I don't see why we should actively phase out Cloud9. I find that it works extremely well, and it allows me to work very efficiently. Could you please explain what limitations are brought by their EULA?

**Comment by notriddle** [comment]
It's a "no commercial use" license. We can use it, since we're neither charging for it nor allowing people to work on closed source software.

But, it means Janitor can't be used internally by proprietary software companies. [Bors-NG], for example, has been getting a lot of external contribution from people who've deployed it at work, and it's a loss that Janitor can't do the same thing. I would love to use Janitor at work, too.

It means Janitor can't become a commercial enterprise itself. I know you don't want to turn Janitor into a startup, but it's disappointing to not even have the option if things go south.

And it means that Janitor can't be distributed by groups like Debian that strictly adhere to the Open Source Definition, because Janitor being closed-source means that it can't be included in a Debian Blu-Ray distribution.

[Bors-NG]: https://bors.tech/

**Comment by jankeromnes** [comment]
Thanks a lot for explaining all the implications of the "no commercial use" license! You make great points, I love the "Janitor at work" idea (side note: we're working with Mozilla's infra team to deploy Janitor at Mozilla), and I agree that it's best to keep all our options open, just in case.

So 100% ok for adding Eclipse Che support, but ideally no phasing out of anything until we have a replacement in place that works great for everyone. Can we have both Eclipse Che and Cloud9, at least for a time, or do you feel this would make our service unreasonably complicated?

And interesting point about Debian distributing Janitor on Blu-Ray. Is that something that could be interesting/valuable at some point?

**Comment by ishitatsuyuki** [comment]
After some amount of research, I've found that Che has a quite gigantic architecture and the integration wouldn't be easy. On the other hand, [Theia](http://www.theia-ide.org/) is solely the IDE component that Che is also going to [adopt](https://github.com/eclipse/che/issues/8322) in the next generation. We can look into integrating it instead, for the meanwhile.

**Comment by Coder206** [comment]
For Theia we should keep an eye on/help out with https://github.com/theia-ide/theia/issues/1405
